priv/perf/apps/load_test/lib/service/faucet.ex

```diff
@@ -45,48 +69,122 @@ defmodule LoadTest.Service.Faucet do
 
-    updated_state = Map.put(state, :utxos, %{state.utxos | token => {change_utxo, change}})
+    next_faucet_utxo = %Utxo{blknum: blknum, txindex: txindex, oindex: 0, amount: change}
+    user_utxo = %Utxo{blknum: blknum, txindex: txindex, oindex: 1, amount: amount}
+
+    updated_state = Map.put(state, :utxos, Map.put(state.utxos, currency, next_faucet_utxo))
 
     {:reply, {:ok, user_utxo}, updated_state}
   end
 
-  defp get_funded_faucet_account(opts) do
-    faucet_private_key = Keyword.fetch!(opts, :faucet_private_key)
-    {:ok, faucet} = Account.new(faucet_private_key)
+  @spec check_sufficient_funds(state(), Utxo.address_binary(), pos_integer()) :: Utxo.t()
+  defp check_sufficient_funds(
+         %{utxos: utxos, faucet_account: faucet_account, fee: fee},
+         currency,
+         amount
+       ) do
+    utxo =
+      case Map.has_key?(utxos, currency) do
+        true ->
+          utxos[currency]
+
+        _ ->
+          get_utxos(faucet_account.addr)
+          |> get_largest_utxo_by_currency(faucet_account, currency)
+      end
+
+    case utxo.amount - amount - fee < 0 do
+      true -> deposit(faucet_account, currency)
+      _ -> utxo
+    end
+  end
+
+  @spec get_utxos(Utxo.address_binary()) :: list()
+  defp get_utxos(address) do
+    {:ok, response} =
+      Api.Account.account_get_utxos(
+        Connection.client(),
+        %Model.AddressBodySchema1{
+          address: Encoding.to_hex(address)
+        }
+      )
+
+    Jason.decode!(response.body)["data"]
+  end
+
+  @spec get_largest_utxo_by_currency(list(), Account.t(), Utxo.address_binary()) :: Utxo.t()
+  defp get_largest_utxo_by_currency([], faucet_account, currency), do: deposit(faucet_account, currency)
+
+  defp get_largest_utxo_by_currency(utxos, faucet_account, currency) do
+    utxos = Enum.filter(utxos, fn utxo -> currency == LoadTest.Utils.Encoding.from_hex(utxo["currency"]) end)
+
+    case length(utxos) do
+      0 -> deposit(faucet_account, currency)
+      _ -> get_largest_utxo(utxos, currency)
+    end
+  end
+
+  @spec get_largest_utxo(list(), Utxo.address_binary()) :: Utxo.t()
+  defp get_largest_utxo(utxos, _currency) do
+    utxo = Enum.max_by(utxos, fn x -> x["amount"] end)
+
+    %Utxo{
+      blknum: utxo["blknum"],
+      txindex: utxo["txindex"],
+      oindex: utxo["oindex"],
+      amount: utxo["amount"]
+    }
+  end
+
+  @spec deposit(Account.t(), Utxo.address_binary()) :: Utxo.t()
+  defp deposit(faucet_account, currency) do
+    Logger.debug("Not enough funds in the faucet, depositing more from the root chain")
 
-    deposit_amount = Keyword.fetch!(opts, :faucet_deposit_wei)
-    deposit_finality_margin = Keyword.fetch!(opts, :deposit_finality_margin)
-    {:ok, deposit_utxo} = Deposit.deposit_from(faucet, deposit_amount, @eth, deposit_finality_margin)
+    {:ok, utxo} =
+      Deposit.deposit_from(
+        faucet_account,
+        Application.fetch_env!(:load_test, :faucet_deposit_wei),
```

Comment:
elixir-omg is following the pattern of passing the configuration top down. which means that a process never access the configuration directly - from the shared resource like application, but the supervisor gives it to them via the arguments to `start_link/1`

Reply:
Hmmm, I agree with the pattern, but then I find that I end up passing these values around everywhere.
I guess that in itself is a code smell...

But are you saying that all config values should be stored in the GenServer state for later use? If not, then where?

Reply:
This is a fairly good example https://github.com/omisego/elixir-omg/blob/master/apps/omg_watcher/lib/omg_watcher/sync_supervisor.ex#L62-L66
The purpose of this is simply to not expose your process to a "shared memory problem". Because `Application` is not pure (there's an underlying ETS), you're allowed to change the configuration in runtime and the consequences of the config change are hard to predict, because they would apply to all processes in the VM, but at different times (exactly when they access it).
So that's why you try to make the process state ... self sufficient.

Also, not saying you should change it now, but it's a fairly good pattern to follow, makes the code a bit more concise as well.